MaxText/layers/linears.py

```diff
@@ -548,16 +551,15 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
 
     if self.config.capacity_factor > 0:
       # token dropping if needed
-      dispatch_mask, combine_mask = self.generate_masks(top_k_indices, softmax_probs)
+      weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)
+      dispatch_mask, combine_mask = self.generate_masks(top_k_indices, weights)
       mask_axes = ("activation_batch", "activation_length", None, None)
       dispatch_mask = nn.with_logical_constraint(dispatch_mask, mask_axes)
       combine_mask = nn.with_logical_constraint(combine_mask, mask_axes)
       loss = self.load_balance_loss(top_k_indices, softmax_probs)
       inputs = nn.with_logical_constraint(inputs, ("activation_batch", "activation_length", "activation_embed"))
       with jax.named_scope("dispatch"):
-        dispatch = self.get_einsum(rhs_mesh_axes=mask_axes)(
-            "BSM,BSEC -> EBCM", inputs, dispatch_mask, precision=matmul_precision
-        )
+        dispatch = jnp.einsum("BSM,BSEC -> EBCM", inputs, dispatch_mask, precision=matmul_precision)
```

Comment:
@ZhiyuLi-goog could you help check if this impact the training performance?

Reply:
regression from 55.0% mfu to 54.5% after changing the 2 ops in to non quantized version:
* per_layer dispatch: from 2.455ms to 3.739ms
* per_layer combine: from 2.086ms to 3.213ms
The regression should be deterministic. The % of tflops from the 2 ops is small.

Recommend to add one more condition for this 2 ops:
```
if self.model_mode == common_types.MODEL_MODE_TRAIN
```